profiling/utils.py

```diff
@@ -0,0 +1,364 @@
+                            input_is_parallel=True,
+                            config=config,
+                            init_method=init_method,
+                        ),
+                    )
+                    for _ in range(self.n_layers // 2)
+                )
+            )
+        else:
+            layers = [
+                nn.Linear(model_dim, model_dim, dtype=self.config.params_dtype)
+                for _ in range(self.n_layers)
+            ]
+        self.layers = nn.ModuleList(layers)
+
+        self.act_fn = nn.ReLU()
+
+    def forward(self, inputs):
+        rep = inputs
+        for i in range(len(self.layers)):
+            rep = self.layers[i](rep)
+
+            # Column and row parallel layers return tuple(output, output_bias).
+            if isinstance(rep, tuple):
+                rep = rep[0]
+
+            if i % 2 == 0:
+                rep = self.act_fn(rep)
+
+        return rep
+
+
+def hook_function_identity(self, inputs, outputs, name):
+    """Hook function that does nothing."""
+    return outputs
+
+
+def hook_function_cpu(self, inputs, outputs, acc, name):
+    """Hook function that dumps to cpu."""
+    rank = mpu.get_tensor_model_parallel_rank()
+
+    # NOTE: See note about output of col and row parallel.
+    _outputs = outputs[0] if isinstance(outputs, tuple) else outputs
+
+    if rank == 0:
+        acc[name] = _outputs.detach().cpu()
+    return outputs
+
+
+def hook_function_gpu(self, inputs, outputs, acc, name):
+    """Hook function that dumps to GPU."""
+    rank = mpu.get_tensor_model_parallel_rank()
+
+    _outputs = outputs[0] if isinstance(outputs, tuple) else outputs
+
+    if rank == 0:
+        acc[name] = _outputs.detach()
+
+    return outputs
+
+
+def hook_function_cpu_gather_scatter(self, inputs, outputs, acc, name):
+    """Hardcoded minimal hook function with gather/scatter.
+    """
+    rank = mpu.get_tensor_model_parallel_rank()
+
+    _outputs = outputs[0] if isinstance(outputs, tuple) else outputs
+    original_shape = _outputs.shape
+
+    def all_gather(tensor, dim=-1):
+        output_list = [
+            torch.empty_like(tensor)
+            for _ in range(mpu.get_tensor_model_parallel_world_size())
+        ]
+        torch.distributed.all_gather(
+            output_list, tensor, group=mpu.get_tensor_model_parallel_group(),
+        )
+        return torch.cat(output_list, dim=dim)
+
+    def scatter(tensor, dim=-1):
+        scatter_list = list(
+            torch.chunk(tensor, mpu.get_tensor_model_parallel_world_size(), dim=dim)
+        )
+        return scatter_list[rank]
+
+    # Determine correct gather/scatter functions.
+    if isinstance(self, ColumnParallelLinear):
+        gather_fn = all_gather
+        scatter_fn = scatter
+
+    elif isinstance(self, RowParallelLinear):
+        gather_fn = lambda x: x
+        scatter_fn = lambda x: x
+
+    else:
+        gather_fn = lambda x: x
+        scatter_fn = lambda x: x
```

Comment:
This option should be `isinstance(self, nn.Linear)` and we have a fallback `else` which raises an implementation error.